David physics 2

drichte2 opened the pull request:
Adds the ability of droplets to independently fall out and be injected into the domain. "nparcelsActiveMax" is the maximum value that "nparcelsActive" can become, which is always less than "nparcelsLocal" by the buffer size. Also fixed a major performance issue due to single precision issues in the nonlinear, iterative droplet solver.

drichte2 committed:
Droplets can now fall out of domain and get removed from pdata. nparcelsActive gets updated and pdata is collapsed with only those still active

drichte2 committed:
Fixed a major performance issue: the Gauss solver for the droplets was reaching max iterations every time when done at single precision. Changing these calculations to double precision speeds up the particle solver 1000x because particles are no longer iterating 1000 times every time. Tests are ongoing for stray particles which can have negative radii

drichte2 committed:
Long winding search for subtle bugs in the droplet nonlinear solver. Needed stronger conditions on when to bail and send out the num100 and num1000 flags. Now the spray injection framework is ready and the code is significantly faster since droplet iterations are no longer maxing out

drichte2 committed:
Cleaned up some things, moved several items back to single precision since they didn't require double

sjsprecious commented on src/cm1.F:
```diff
@@ -2259,11 +2259,12 @@ program cm1
 #ifdef MPI
-      padding = max(pdata_buffer*nparcels,100.0)
-      nparcelsLocal = ceiling((nparcels*1.0/numprocs)+padding)
+      padding = max(pdata_buffer*nparcelsActiveMax,100.0)
```
By making this change, the value of `padding` is now independent of the `nparcels` value. I think a sanity check is needed to make sure that `nparcels/numprocs` is always smaller than `nparcelsActiveMax`.

drichte2 replied:
Good point. The padding should actually be a percentage of `nparcelsActiveMax`, since otherwise the padding would need to be adjusted throughout the simulation (which doesn't seem ideal/possible). So maybe I can change to `nparcelsActiveMax/numprocs` instead?

johnmauff replied:
I am also not comfortable with this change both for the reason that Jian mentioned and also for the fact that there are now two places where the number of droplets is set.  Once in the namelist and then once in this file.  That is a recipe for confusion.

sjsprecious replied:
Thanks @drichte2 . It is up to you to decide the calculation of `padding`. My point is that we need to check the value of `nparcels` here. For example, if I set `nparcels = 1 billion`, each MPI rank will have more than `nparcelsActiveMax` droplets and ideally the program should stop with an error message. This is missing from the current implementation.

drichte2 replied:
I just added a quick check to see if `nparcels` exceeds `nparcelsActiveMax`. @johnmauff , it's not setting the same thing twice. `nparcels` is the number of droplets to start the simulation with. `nparcelsActiveMax` is a maximum upper limit on the number of parcels that can ever be active at the same time. This would be something that should only be changed with architecture/memory limitations

johnmauff replied:
@drichte2  Historically nparcels was the total number of particles.  I think the current use is a bit confusing and error prone.   Maybe we need to have two namelist variables:  
 
nparcelsInit:  Initial number of particles
nparcelsMax:  Maximum number of particles.  

Might want to get @george-bryan input as we are proposing a namelist change.

drichte2 replied:
I agree with the last part -- I was trying to do this in a way that didn't mess with the namelist file. I'd be fine adding it at some point, but didn't want to mess up everyone's namelists

sjsprecious commented on src/Makefile:
```diff
@@ -156,7 +156,7 @@ endif
     # replace -xHost with -march=core-avx2 to make sure the same code builds on
     # both Intel and AMD CPUs 
-    OPTS         += -g -O2 -march=core-avx2 -ip -assume byterecl -fp-model precise -ftz -init=snan,arrays -traceback -no-fma
+    OPTS         += -g -O2 -xHost -ip -assume byterecl -fp-model precise -ftz -init=snan,arrays -traceback -no-fma
```
I think this flag won't work on Gust's CPU. Better to revert this change.

drichte2 replied:
Yes, this was a mistake since it doesn't work on my local machine. Will change for the pull request

sjsprecious commented on src/droplet.F:
```diff
@@ -126,9 +133,20 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
       integer :: tmp_flag
 #endif
 
+      double precision :: t_s,t_f,t_findloc,t_interp,t_BEint,t_feed,t_s1,t_f1,t_BEprep,t_radsolver,t_gauss,t_LV,t_neighbor
```
Do we need to keep these timers during our ASD runs or are they just for your own diagnostic?

drichte2 replied:
No, I meant to take these out. I have removed them from the pull request

sjsprecious commented on src/droplet.F:
```diff
@@ -126,9 +133,20 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+
+      t_s = mpi_wtime()
+
+      !!! DHR: From what I can tell this is useless -- it only resets seed for the intrinsic random_number routine
```
I agree and I thought you would remove these calls and take the `randomNumbers` from the `init3d` subroutine to pull new random numbers.

drichte2 replied:
Okay I have removed this

sjsprecious commented on src/droplet.F:
```diff
@@ -302,9 +319,27 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
     !$acc loop gang vector
 #endif
 
+    t_f = mpi_wtime()
+    if (myid==0) write(*,*) 'TIME prep:',t_f-t_s
+
+    t_s = mpi_wtime()
+
+    t_findloc = 0.0
+    t_interp = 0.0
+    t_BEint = 0.0
+    t_feed = 0.0
+
+    t_BEprep = 0.0
+    t_radsolver = 0.0
+    t_gauss = 0.0
+    t_LV = 0.0
+    t_neighbor = 0.0
+
```
These changes will break the GPU code. Need to be moved up to line 311.

drichte2 replied:
These were all for my timing tests, I have removed them

sjsprecious commented on src/droplet.F:
```diff
@@ -445,8 +468,73 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+
+   deallocate(holes_fallout_ind)
+
+   call mpi_allreduce(nparcelsActive,np_tmp,1,mpi_int,mpi_sum,mpi_comm_world,ierr)
```
why do we need an `mpi_allreduce` call here?

drichte2 replied:
I have moved this down below. At the moment it's just for a writeout statement that lets me know how many particles are currently in the system (variable `np_tmp`). This can be removed later

sjsprecious commented on src/droplet.F:
```diff
@@ -527,23 +617,49 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+        s10_xym = s10_xym + s10(i,j)
+     end do
+     end do
+     call mpi_allreduce(mpi_in_place,s10_xym,1,mpi_real,mpi_sum,mpi_comm_world,ierr)
```
Again, why do we need an `mpi_allreduce` call here?

drichte2 replied:
All of the processors need to know the value of `s10_xym`. This is the 10-meter wind speed that goes into the droplet injection routine and determines how many droplets need to be produced.

sjsprecious replied:
But this means all the processors will know the maximum 10-m wind speed over the whole domain. Should it be enough for each processor to know the 10-m wind speed of its own sub-domain?

drichte2 replied:
Strictly speaking, it should really be the domain-wide average. In practice, it probably wouldn't make a big difference if we did it locally, but this would introduce a dependence on `numprocs` since different processor numbers would lead to different areas that this averaging is being done over.

sjsprecious replied:
thanks for the clarification.

sjsprecious commented on src/droplet.F:
```diff
@@ -645,34 +764,48 @@ subroutine BE_integration(dt,np,pdata,pdata_locind,xf,yf,x3d,y3d, &
+      if (flag==1)  then
+
+       num1000 = num1000+1 
+
```
The `if` checks for `rt_zeros` are removed, is it intended or not?

drichte2 replied:
Yes, this was intended. The checks on `rt_zeros` are now above, and they set `flag = 1` which triggers this code.

sjsprecious commented on src/droplet.F:
```diff
@@ -689,11 +822,12 @@ subroutine BE_integration(dt,np,pdata,pdata_locind,xf,yf,x3d,y3d, &
+         !z3d = bot + (bot-z3d)
+         !pdata(np,prvpz) = -pdata(np,prvpz)
          pdata_locind(np,3) = undefined_index
+         num_fallout = num_fallout + 1
```
For my own reference: `num_fallout` needs to be put on the `reduction` clause for GPU code.

sjsprecious commented on src/droplet.F:
```diff
@@ -2270,6 +2419,196 @@ subroutine ie_vrt_nd(diffnorm,tempr,tempt,rt_output,dt_nondim,taup_scale,prsval,
+      do np=nparcelsActive+1,nparcelsActive+my_reintro
+
+         rand = getRandomReal(randomNumbers)
+         a = totdrops*getRandomReal(randomNumbers)
```
should it be `a = totdrops * rand`?

drichte2 replied:
Yes, my mistake

sjsprecious commented on src/droplet.F:
```diff
@@ -2270,6 +2419,196 @@ subroutine ie_vrt_nd(diffnorm,tempr,tempt,rt_output,dt_nondim,taup_scale,prsval,
+         rand = getRandomReal(randomNumbers)
+         a = totdrops*getRandomReal(randomNumbers)
+
+         do i = 1,100
+            if ((a .gt. dFssum(i)) .and. (a .le. dFssum(i+1))) then
+               rand = getRandomReal(randomNumbers)
+               rad_init = binsdata(i) + rand*(binsdata(i+1)-binsdata(i))
+            endif
+         enddo
```
Can you add some comments about what this loop is doing here?

drichte2 replied:
I have added some comments. Let me know if they don't explain well

sjsprecious reviewed (changes requested):
Thanks @drichte2 for submitting this PR. I have a few questions and hope you can clarify them before I can work on the fix for the GPU code.

sjsprecious commented:
@drichte2 : also how did you verify the correctness of your new code?

johnmauff commented on src/droplet.F:
```diff
@@ -445,8 +468,73 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+    if (myid==0) write(*,*) 'TIME feed:',t_feed
+
+    call mpi_allreduce(mpi_in_place,num100,1,mpi_int,mpi_sum,mpi_comm_world,ierr)
+    call mpi_allreduce(mpi_in_place,num1000,1,mpi_int,mpi_sum,mpi_comm_world,ierr)
```
Why are we performing two calls to MPI_Allreduce? A better solution would be to perform a single MPI_Allreduce call on a vector of length two.

drichte2 replied:
Changed

johnmauff commented on src/init3d.F:
```diff
@@ -423,7 +423,8 @@ subroutine init3d(xh,rxh,uh,ruh,xf,rxf,uf,ruf,yh,vh,rvh,yf,vf,rvf,  &
           rand = getRandomReal(randomNumbers)
-          tmpz = rand*maxz
+          !tmpz = rand*maxz
+          tmpz = rand*8.0
```
Could you explain why the variable maxz was replaced with a magic number 8?

drichte2 replied:
This is the maximum height that the droplets will be "injected". Problem-specific, and has to do with things like the (unresolved) wave height, etc. Using `maxz` here is like saying that a new droplet could magically appear anywhere in the domain, when in reality it should start near the bottom.

johnmauff replied:
Ok that makes sense.  We probably need to add this as a namelist parameter.  The current use of a number that is not documented makes the code more fragile than necessary.

drichte2 replied:
Agreed -- see my other comment about adding things to the namelist. That would make sense, but didn't want to mess with that part

johnmauff commented on src/misclibs.F:
```diff
@@ -2986,7 +2986,7 @@ subroutine calccfl(nstat,rstat,dt,acfl,uh,vh,mh,ua,va,wa,writeit)
 !!!      cflmax = fmax
 
-      print *,'calccfl: fmax is: ',fmax
+!      print *,'calccfl: fmax is: ',fmax
```
Lets just delete this print statement and the next one on line 3170.t

drichte2 replied:
Done

drichte2 commented:
Correctness was done by ensuring the droplet solver for a fixed background value-- this was how I had done this the first time.

drichte2 committed:
Made some changes for the pull request

sjsprecious commented:
> Correctness was done by ensuring the droplet solver for a fixed background value-- this was how I had done this the first time.

Thanks @drichte2 . It is still unclear to me how to verify the correctness quantitatively, like which statistic output you are looking at. We should have an offline discussion about this in detail.

drichte2 committed:
Added a check to make sure that nparcels doesn't exceed nparcelsActiveMax

drichte2 committed:
Changed Makefile

drichte2 committed:
A few more minor changes

drichte2 committed:
Some changes to only inject droplets over some elapsed time

drichte2 committed:
Fixed a conflict with a change made to the repository summing num100 and num1000

drichte2 committed:
In fixing feedback projection, added several terms to the droplet diagnostics

drichte2 committed:
Fixed a critical bug in the feedback calculation

drichte2 committed:
Another adjustment; things working; now comparing to the GRL paper

drichte2 committed:
Cleaned things up

sjsprecious committed:
Merge branch 'gpu-opt' into david_physics_2

resolve Conflicts:
	src/input.F

sjsprecious committed:
Update namelist variables nparcelsInit and nparcelsMax
Update nparcelsLocalActive as name convention for each MPI rank

	modified:   run/namelist.restA
	modified:   run/namelist.restB
	modified:   run/namelist.restC
	modified:   src/Makefile
	modified:   src/cm1.F
	modified:   src/comm_droplet.F
	modified:   src/constants.F
	modified:   src/droplet.F
	modified:   src/init3d.F
	modified:   src/input.F
	modified:   src/param.F
	modified:   src/parcel.F
	modified:   src/restart.F

sjsprecious committed:
Add injectHMax namelist variable for droplet injection
Add flag to switch fallout condition for droplets

	modified:   run/namelist.restA
	modified:   run/namelist.restB
	modified:   run/namelist.restC
	modified:   src/droplet.F
	modified:   src/init3d.F
	modified:   src/input.F
	modified:   src/param.F
	modified:   src/restart.F

sjsprecious committed:
fix a typo for variable type

	modified:   input.F

sjsprecious committed:
Print nparcelsMax

	modified:   param.F

sjsprecious committed:
Update namelist.restB for restart droplet number
(set local_seed to myid to obtain this droplet number)

	modified:   namelist.restB

johnmauff commented on src/param.F:
```diff
@@ -297,7 +297,7 @@ subroutine param(namelist_input,                                  &
       call bcast_int(pdtra)
       call bcast_int(iprcl)
-      call bcast_int(nparcels) ! DEVICE
+      call bcast_int(nparcelsInit) ! DEVICE
```
@sjsprecious Where is the corresponding bcast_int() call for nparcelsMax?

sjsprecious replied:
@johnmauff thanks for catching this! Just added it.

johnmauff commented on src/cm1.F:
```diff
@@ -2259,15 +2259,18 @@ program cm1
       reqt = 0
 
+      if ( nparcelsInit .gt. nparcelsMax ) then
+          if ( myid .eq. 0 ) print *,' WARNING: nparcelsInit exceeds allowed max. Check nparcelsMax. Exiting...'
```
@sjsprecious  I am concerned that you have removed the necessary padding for setting the value for nparcelsLocal.  How do you prevent code failure if the number of incoming particles exceeds nparcelsLocal.

sjsprecious replied:
@johnmauff with the new namelist variable `nparcelsMax`, `nparcelsLocal` is now calculated as `nparcelsMax/numproc` and I think we do not need the `padding` variable now. In the `init3d.F` subroutine, there is a check if the number of incoming particles exceeds `nparcelsLocal` and an error will issue accordingly.

johnmauff replied:
@sjsprecious I think the padding still needs to be present.  Right now the code could fail unpredictably later in the simulation if the number of incoming droplets is sufficiently large.  Another option to test this functionality would be to set nparcelsMax==nparcelsInit.  The code should then fail for the first MPI rank that receives more incoming droplets than depart.

sjsprecious replied:
@johnmauff thanks for the detailed explanation. That makes sense. However, if we use the variable `padding` to increase the size of `nparcelsLocal`, how could we ensure that the total droplets after injection will not exceed `nparcelsMax` at each time step?

sjsprecious committed:
add bcast call for nparcelsMax and injectHMax

	modified:   param.F

johnmauff commented:
@sjsprecious, @drichte2  @sherimickelson I just took a close look at these changes and I have an increasingly large number of concerns.  For example, the droplet code is mostly missing any comments, there has been code added that will prevent the code from compiling without defining MPI and there is still no clarity regarding how to set various maximum # of particles.  There is also missing OpenACC support for several newly added loops.  I am concerned that allowing this PR will immediately break a number of execution modes for CM1 that currently work.  Ideally, I would like to run some performance measurements on David's new additions, but I don't understand how I can configure the code so that it runs on a constant particle # to allow comparison with the ISC paper results.  I wonder if we need to discuss these code changes synchronously instead of through GitHub.

sjsprecious commented:
Thanks @johnmauff . I agree that we may need another meeting to address these issues. I also find that with the changes here, the restart run does not agree with the initial run anymore (though the difference is not that large, e.g., total droplet number).

sjsprecious committed:
turn on the droplet injection only when use_wall_bc is FALSE, now conserving the total droplet number
a few minor fixes

	modified:   droplet.F
	modified:   init3d.F
	modified:   restart.F

sjsprecious commented:
@johnmauff In terms of conserving the total droplet number, I just pushed a new change and now by setting `use_wall_bc = .true.`, we will have a constant droplet number through the simulation.

drichte2 committed:
Fixed loop bounds in statstics

drichte2 committed:
Resolved merge conflicts after Jian added the nparcelsMax and nparcelsLocalActive

drichte2 committed:
Added comments and IFDEF statements for compilation without MPI. Also cleaned up a few other items

sjsprecious committed:
add back the buffer to pdata allocation to account for the droplet communication between MPI ranks

	modified:   cm1.F
	modified:   constants.F
	modified:   init3d.F

drichte2 committed:
Change minimum buffer size to 500

drichte2 committed:
Added random_droplet parameter at top of droplets.F to test restart. Changes to purely deterministic droplet injection

sjsprecious committed:
Update namelist for initial and restart run tests

	modified:   namelist.restA
	modified:   namelist.restB
	modified:   namelist.restC

sjsprecious committed:
Revise the calculation form of seed value
Reset the information for dead droplets
Add more comments

	modified:   run/namelist.restB
	modified:   src/droplet.F

sjsprecious commented on src/droplet.F:
```diff
@@ -400,53 +416,72 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+    numvec(1) = num100
+    numvec(2) = num1000
+
+#ifdef MPI
+
+    if ( myid == 0 ) then
+       call mpi_reduce(mpi_in_place,numvec,2,mpi_int,mpi_sum,0,mpi_comm_world,ierr)
+    else
+       call mpi_reduce(numvec,numvec,2,mpi_int,mpi_sum,0,mpi_comm_world,ierr)
+    end if
+
+#endif
+
+    if (myid==0) write(*,*) 'num100,num1000 = ',numvec(1),numvec(2)
```
Are these for debug purpose only or do we need to include them in a production run?

drichte2 replied:
I'll remove the print statement in the next PR, but the calculations need to happen. I need to send them to the droplet_diag

sjsprecious commented on src/droplet.F:
```diff
@@ -515,35 +550,61 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+#ifdef MPI
+      call mpi_allreduce(nparcelsLocalActive,np_tmp,1,mpi_int,mpi_sum,mpi_comm_world,ierr)
+#endif
+      if (myid==0) then 
+         write(*,'(a8,3i)')  'DHR1:',myid,np_tmp,nparcelsLocalActive
+      end if
```
Do we need to keep this print statement?

drichte2 replied:
No, not in the long run. I'll remove it for the next PR

sjsprecious commented on src/droplet.F:
```diff
@@ -515,35 +550,61 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
+      call mpi_allreduce(mpi_in_place,s10_xym,1,mpi_real,mpi_sum,mpi_comm_world,ierr)
+#endif
+      s10_xym = s10_xym/real(nx*ny)
+      if (myid==0) write(*,*) 'DHR U10',s10_xym
```
Do we need to keep this print statement?

drichte2 replied:
Same -- not in the long run, but it's helpful now as I set up the ASD case. I'll remove for the next PR

drichte2 committed:
Merge branch 'gpu-opt' into david_physics_2

Merge gpu-opt into david_physics_2 after the droplet injection is now functional

drichte2 committed:
Added prt to use list

drichte2 committed:
Cleaned up merge conflicts when pulling the merged gpu-opt+david_physics_2 after droplet injection was added

drichte2 committed:
Cleaned up two small bugs

sjsprecious committed:
Add a missing statement for use_wall_bc
Leave a comment about the issue of using "time_dropinject"

	modified:   droplet.F

sjsprecious commented:
Somehow I could not revert some commits in this PR so I end up with opening a separate PR for all the approved changes (available at https://github.com/george-bryan/CM1/pull/100) and closing this PR without merging.
